trl/trainer/dpo_trainer.py

```diff
@@ -1284,190 +1162,148 @@ def dpo_loss(
+            # Mask the prompt but not the completion for the loss
+            loss_mask = torch.cat(
+                (torch.zeros_like(prompt_attention_mask), completion_attention_mask),
+                dim=1,
+            )
 
-        if self.loss_type == "ipo":
-            all_logps = all_logps / size_completion
+            # Flush left to reduce the memory usage
+            # [[0, 0, x, x, x, x],  ->  [[x, x, x, x],
+            #  [0, x, x, x, 0, 0]]       [x, x, x, 0]]
+            for i in range(attention_mask.size(0)):
+                first_one_idx = torch.nonzero(attention_mask[i])[0].item()
+                input_ids[i] = torch.roll(input_ids[i], shifts=-first_one_idx)
+                attention_mask[i] = torch.roll(attention_mask[i], shifts=-first_one_idx)
+                loss_mask[i] = torch.roll(loss_mask[i], shifts=-first_one_idx)
+
+            # Get the first column idx that is all zeros and remove every column after that
+            empty_cols = torch.sum(attention_mask, dim=0) == 0
+            first_empty_col = torch.nonzero(empty_cols)[0].item() if empty_cols.any() else attention_mask.size(1) + 1
+            input_ids = input_ids[:, : first_empty_col - 1]
+            attention_mask = attention_mask[:, : first_empty_col - 1]
+            loss_mask = loss_mask[:, : first_empty_col - 1]
+
+            # Truncate left
+            if self.args.max_length is not None:
+                input_ids = input_ids[:, -self.args.max_length :]
+                attention_mask = attention_mask[:, -self.args.max_length :]
+                loss_mask = loss_mask[:, -self.args.max_length :]
+
+            outputs = model(input_ids=input_ids, attention_mask=attention_mask, **model_kwargs)
+
+            # Offset the logits by one to align with the labels
+            logits = outputs.logits[:, :-1, :]
+            labels = input_ids[:, 1:].clone()
+            loss_mask = loss_mask[:, 1:].bool()
+
+        if logits.shape[:2] != labels.shape[:2]:
+            # for llava, the returned logits include the image tokens (placed before the text tokens)
+            seq_len = labels.shape[1]
+            logits = logits[:, -seq_len:]
+
+        # Compute the log probabilities of the labels
+        labels[~loss_mask] = 0  # dummy token; we'll ignore the losses on these tokens later
+        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)).squeeze(2)
+        per_token_logps[~loss_mask] = 0
+        all_logps = per_token_logps.sum(-1)
+
+        output = {}
 
-        policy_weights = None
         if self.use_weighting:
-            chosen_weights = all_weights[:len_chosen]
-            rejected_weights = all_weights[len_chosen:]
-            policy_weights = torch.clamp(torch.exp(chosen_weights + rejected_weights), max=1)
+            with torch.no_grad():
+                # Eq (2) of the WPO paper: https://huggingface.co/papers/2406.11827
```

Comment:
Is this weighting applied for vanilla DPO as well or is it excluded since the loss doesn't depend on the `policy_weights` attribute?

Reply:
arf sorry i was being dumb - i missed the `self.use_weighting` line above and thought we were computing `policy_weights` by default. ignore my question :)